[Merged by Bors] - v1.1.6 Fork Choice changes

realbigsean opened the pull request:
## Issue Addressed

Resolves: https://github.com/sigp/lighthouse/issues/2741
Includes: https://github.com/sigp/lighthouse/pull/2853 so that we can get ssz static tests passing here on v1.1.6. If we want to merge that first, we can make this diff slightly smaller

## Proposed Changes

- Changes the `justified_epoch` and `finalized_epoch` in the `ProtoArrayNode` each to an `Option<Checkpoint>`. The `Option` is necessary only for the migration, so not ideal. But does allow us to add a default logic to `None` on these fields during the database migration.
- Adds a database migration from a legacy fork choice struct to the new one, search for all necessary block roots in fork choice by iterating through blocks in the db.
- updates related to https://github.com/ethereum/consensus-specs/pull/2727
  -  We will have to update the persisted forkchoice to make sure the justified checkpoint stored is correct according to the updated fork choice logic. This boils down to setting the forkchoice store's justified checkpoint to the justified checkpoint of the block that advanced the finalized checkpoint to the current one. 
  - AFAICT there's no migration steps necessary for the update to allow applying attestations from prior blocks, but would appreciate confirmation on that
- I updated the consensus spec tests to v1.1.6 here, but they will fail until we also implement the proposer score boost updates. I confirmed that the previously failing scenario `new_finalized_slot_is_justified_checkpoint_ancestor` will now pass after the boost updates, but haven't confirmed _all_ tests will pass because I just quickly stubbed out the proposer boost test scenario formatting.
- This PR now also includes proposer boosting https://github.com/ethereum/consensus-specs/pull/2730

## Additional Info
I realized checking justified and finalized roots in fork choice makes it more likely that we trigger this bug: https://github.com/ethereum/consensus-specs/pull/2727

It's possible the combination of justified checkpoint and finalized checkpoint in the forkchoice store is different from in any block in fork choice. So when trying to startup our store's justified checkpoint seems invalid to the rest of fork choice (but it should be valid). When this happens we get an `InvalidBestNode` error and fail to start up. So I'm including that bugfix in this branch.

Todo:

- [x] Fix fork choice tests
- [x] Self review
- [x] Add fix for https://github.com/ethereum/consensus-specs/pull/2727
- [x] Rebase onto Kintusgi 
- [x] Fix `num_active_validators` calculation as @michaelsproul pointed out
- [x] Clean up db migrations

realbigsean committed:
move `Legacy` structs to `schema_change.rs`

realbigsean committed:
update justified epoch to checkpoint in fork choice

realbigsean committed:
Initial migration implementation

realbigsean committed:
cleanup

realbigsean committed:
More cleanup, some bug fixes

realbigsean committed:
update/refactor to include adding the finalized root to the fork choice

realbigsean committed:
make sure `node_mutator` is applied to the finalized block

realbigsean committed:
Don't update common ancestors

realbigsean committed:
Correctly find heads

realbigsean committed:
add some comments, use correct index in `node_mutator`

realbigsean committed:
small cleanup

realbigsean committed:
Update comments

realbigsean committed:
Make `proto_array.finalized_checkpoint` and `proto_array.justified_checkpoint` required fields (remove `Option`)

realbigsean committed:
Improve `Option` handling in fork choice

realbigsean committed:
Small test refactors

realbigsean committed:
Fork choice test updates

realbigsean committed:
fix ffg_updates tests

realbigsean committed:
fix no_votes tests

realbigsean committed:
fix votes tests

realbigsean committed:
self review updates to schema migration code

realbigsean committed:
move new schema migration code to a different mod

realbigsean committed:
fix typos and delete unused code

realbigsean committed:
delete dbgs

realbigsean committed:
small cleanup

realbigsean committed:
remove semicolons from macros

realbigsean committed:
cargo fmt

realbigsean committed:
add a couple of logs and prune prior to schema migration

realbigsean committed:
add some logs

realbigsean committed:
De dupe correctly

realbigsean committed:
Revert "add some logs"

This reverts commit fc14c446750845d59a659cacd4d501a52eab3ca7.

realbigsean committed:
Revert 8c96fc5136e7ad951b0cee849a28452be6eb1599

realbigsean committed:
fix merge

realbigsean committed:
fallback to initializing fork choice from an anchor state

realbigsean committed:
fix fallback in migration

realbigsean committed:
cargo fmt

realbigsean committed:
Spec Atomicity bug updates

realbigsean committed:
No longer skip `new_finalized_slot_is_justified_checkpoint_ancestor` EF test scenario, increment tests to 1.1.6

realbigsean committed:
Remove unused getters

realbigsean committed:
Separate kintsugi migration from the roots migration

realbigsean committed:
log out errors before falling back in the database migration

realbigsean committed:
self-review

realbigsean committed:
self-review

michaelsproul committed:
Start on proposer boosting

realbigsean committed:
Implement proposer boost

realbigsean commented:
Still failing an EF test.  This implementation differs from the spec a bit in how this issue is handled: https://github.com/ethereum/consensus-specs/issues/2757

So marking this as blocked for now

michaelsproul commented on consensus/proto_array/src/proto_array.rs:
```diff
@@ -142,11 +125,24 @@ impl ProtoArray {
+                node_delta -= self.previous_proposer_boost.score as i64;
+            }
+            if proposer_boost_root != Hash256::zero() && proposer_boost_root == node.root {
+                let num_validators = new_balances.len() as u64;
```
I'm looking at some other fork choice code and I noticed that the balances arrays in fork choice can contain entries for inactive validators, which would make this calculation incorrect. Fork choice uses this function to compute balances:

https://github.com/sigp/lighthouse/blob/fff01b24ddedcd54486e374460855ca20d3dd232/beacon_node/beacon_chain/src/beacon_fork_choice_store.rs#L39-L55

I'm not _certain_ that the balances array passed in here comes from that function, but wanted to flag that as something we should check

realbigsean replied:
Yes, you're right and we do use that balances array here. It's set in the fork choice store in `set_justified_checkpoint`. Nice catch!

But it also looks like when there's a cache miss in `set_justified_checkpoint`... we are getting balances from `state.balances` where we have exact balances. And I don't think we check whether validators are active at that point. So there seem to be a couple issues here :sweat:

michaelsproul replied:
Ooft! Ok!

I think once we have the correct effective balances here we _should_ be able to just filter out the 0 balances? So the number of active validators would be `new_balances.iter().filter(|b| b != 0).count()` and the total active balance would be `new_balances.iter().filter(|b| b != 0).sum()`. If we wanted to be super efficient we could compute the sum and the count in one iteration using a `fold` (or a loop).

realbigsean replied:
So I guess this is a question of whether we can have active validators with an effective balance of zero? I don't think it's currently possible and it doesn't make sense intuitively, so I'm good with this route.

paulhauner commented:
Github closed this automatically, not me. Sorry!

realbigsean committed:
add `SkippedKnownFailure` where necessary

realbigsean committed:
v1.1.6 field/method renaming

realbigsean committed:
1.57.0 lints

realbigsean committed:
PR updates

realbigsean committed:
cargo fmt

realbigsean committed:
add back the `Lockfile::file`

realbigsean committed:
`proposer_boost_root` migration

realbigsean committed:
Update some comments

realbigsean committed:
only box `QueuedBlock::block` rather than `QueuedBlock`

realbigsean committed:
deserialize unused fields in EF tests

realbigsean committed:
fix 5->7 db migration

realbigsean committed:
fix failing test

michaelsproul committed:
Use superstruct for PersistedForkChoice

michaelsproul commented:
I've opened a PR to this PR here that starts using `superstruct` in the migration process: https://github.com/realbigsean/lighthouse/pull/18

realbigsean committed:
Update beacon_node/beacon_chain/src/schema_change/README.md

realbigsean committed:
Update beacon_node/beacon_chain/src/schema_change/README.md

realbigsean committed:
Merge pull request #18 from michaelsproul/superstruct-migrate

Use superstruct for PersistedForkChoice

realbigsean committed:
Filter inactive validators from the count in calculating the proposer boost

realbigsean committed:
fix comment

realbigsean committed:
cleanup db migrations

realbigsean committed:
fix proposer score calculation

realbigsean committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into justified-checkpoint-root

realbigsean committed:
update variable names

realbigsean committed:
Fix naming

realbigsean committed:
fix feeRecipient naming in tests

realbigsean committed:
fix compile err

realbigsean committed:
Ignore access to files for merge transition EF tests

realbigsean commented on beacon_node/beacon_chain/src/beacon_chain.rs:
```diff
@@ -2443,11 +2447,18 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
         {
             let _fork_choice_block_timer =
                 metrics::start_timer(&metrics::FORK_CHOICE_PROCESS_BLOCK_TIMES);
+            // FIXME(boost): consider using observation delay
```
```suggestion
```

realbigsean committed:
Update beacon_node/beacon_chain/src/beacon_chain.rs

realbigsean commented on common/slot_clock/src/lib.rs:
```diff
@@ -82,24 +83,32 @@ pub trait SlotClock: Send + Sync + Sized + Clone {
+        self.slot_duration() * 2 / INTERVALS_PER_SLOT as u32
+    }
+
+    fn seconds_from_current_slot_start(&self, seconds_per_slot: u64) -> Option<Duration> {
```
```suggestion
    /// Returns the `Duration` since the start of the current `Slot`. Useful in determining whether to apply proposer boosts.
    fn seconds_from_current_slot_start(&self, seconds_per_slot: u64) -> Option<Duration> {
```

realbigsean committed:
Update common/slot_clock/src/lib.rs

realbigsean committed:
full path in ignored ef test file

realbigsean committed:
Merge branch 'justified-checkpoint-root' of https://github.com/realbigsean/lighthouse into justified-checkpoint-root

realbigsean committed:
fix accessed files script

realbigsean commented:
The most recent CI failure was in the simulator, and it passed when I ran it again locally, so not sure there's really an issue. Might be simulator inconsistency. So I restarted the workflow and I think this one's ready for a review.

michaelsproul commented on beacon_node/beacon_chain/src/schema_change.rs:
```diff
@@ -95,22 +100,60 @@ pub fn migrate_schema<T: BeaconChainTypes>(
+                )?;
             }
 
             db.store_schema_version(to)?;
```
I know the existing migrations don't do this, but it would be nice to use a single atomic operation for these modifications so that if a migration crashes half-way then it doesn't leave the database in a corrupt state (e.g. v7 fork choice but with the stored version stating v6).

michaelsproul commented on beacon_node/beacon_chain/src/schema_change/migration_schema_v7.rs:
```diff
@@ -0,0 +1,331 @@
+    let mut iter = std::iter::once(Ok((head_root, head_slot)))
+        .chain(BlockRootsIterator::from_block(db, head_root).map_err(|e| format!("{:?}", e))?);
+    let mut roots_by_epoch = HashMap::new();
+    for epoch in relevant_epochs.into_iter() {
```
```suggestion
    for epoch in relevant_epochs {
```

michaelsproul commented on beacon_node/beacon_chain/src/schema_change/migration_schema_v7.rs:
```diff
@@ -0,0 +1,331 @@
+
+/// Applies a mutator to every node in a chain for descendants from the `finalized_root`, starting
+/// with the node at the given `head_index`.
+fn apply_to_chain_of_descendants<F>(
```
It seems a bit strange that this is named `apply_to_chain_of_descendants` rather than _ancestors_, seeing as we're iterating backwards from a head?

michaelsproul commented on consensus/fork_choice/src/fork_choice_store.rs:
```diff
@@ -19,7 +19,7 @@ use types::{BeaconBlock, BeaconState, Checkpoint, EthSpec, Hash256, Slot};
     type Error;
 
-    /// Returns the last value passed to `Self::update_time`.
+    /// Returns the slot of the last value passed to `Self::set_current_slot`.
```
```suggestion
    /// Returns the last value passed to `Self::set_current_slot`.
```

This was my bad

michaelsproul commented on consensus/proto_array/src/proto_array.rs:
```diff
@@ -142,11 +125,36 @@ impl ProtoArray {
+                let committee_size = num_validators / E::slots_per_epoch();
+                let committee_weight = committee_size * average_balance;
+                proposer_boost_score =
+                    (committee_weight * E::default_spec().proposer_score_boost) / 100;
```
I don't think we should use `E::default_spec()` here, as it obscures potential differences between network configs. Have you looked at how hard it would be to plumb a `ChainSpec` through to this point?

michaelsproul replied:
I think it pays to be defensive and use safe arith here. I think this crate mostly uses `checked_` methods with `DeltaOverflow` explicit errors, but we could also bring in the `safe_arith` crate and add a `From<safe_arith::Error>` impl

michaelsproul commented on testing/ef_tests/src/cases/fork_choice.rs:
```diff
@@ -147,16 +145,13 @@ impl<E: EthSpec> Case for ForkChoiceTest<E> {
-        //
-        // We should eventually solve the above issue and remove this `SkippedKnownFailure`.
-        if self.description == "new_finalized_slot_is_justified_checkpoint_ancestor"
+        // https://github.com/ethereum/consensus-specs/pull/2753
```
```suggestion
        // https://github.com/ethereum/consensus-specs/pull/2760
```

michaelsproul reviewed (commented):
This is looking really solid, I reckon we merge to `unstable` ASAP and continue testing it on our infra so it can go into the next release.

realbigsean committed:
Update testing/ef_tests/src/cases/fork_choice.rs

Co-authored-by: Michael Sproul <michael@sigmaprime.io>

realbigsean committed:
Update consensus/fork_choice/src/fork_choice_store.rs

Co-authored-by: Michael Sproul <michael@sigmaprime.io>

realbigsean committed:
Update beacon_node/beacon_chain/src/schema_change/migration_schema_v7.rs

Co-authored-by: Michael Sproul <michael@sigmaprime.io>

realbigsean committed:
PR feedback

realbigsean committed:
Make schema version db update atomic with other operations in the latest migrations

realbigsean committed:
Merge branch 'justified-checkpoint-root' of https://github.com/realbigsean/lighthouse into justified-checkpoint-root

realbigsean commented on beacon_node/beacon_chain/src/schema_change/migration_schema_v7.rs:
```diff
@@ -0,0 +1,331 @@
+    Ok(roots_by_epoch)
+}
+
+/// Applies a mutator to every node in a chain, starting from the node at the given given
```
```suggestion
/// Applies a mutator to every node in a chain, starting from the node at the given
```

realbigsean committed:
Update beacon_node/beacon_chain/src/schema_change/migration_schema_v7.rs

realbigsean commented:
Ok addressed all comments! Here is the diff: https://github.com/sigp/lighthouse/compare/87e1c61c0fe06e0a6ef73bf4123fd57f88742b60..00a81e025d79c3d921ff365f934012b456599746

michaelsproul reviewed (approved):
Going to run this on all the Prater nodes, and can merge so long as there are no fatal issues.

michaelsproul committed:
Make proposer boosting optional

michaelsproul commented:
I've pushed a commit to make reading the proposer boost from the config optional, which should allow us to merge to unstable (i.e. we can be compatible with existing devnet configs a la https://github.com/sigp/lighthouse/pull/2859). Proposer boosting is enabled on Pyrmont & Prater, and disabled on mainnet until we push a new commit (and release) to enable it.

It would be nice to enable and disable it via a CLI flag but this would require restructuring the CLI config parsing to be able to mutate the `ChainSpec` (currently it's immutable). We can always enable it with a custom network dir (`--testnet-dir`), which could be handy for testing.

michaelsproul committed:
Merge remote-tracking branch 'origin/unstable' into justified-checkpoint-root

michaelsproul commented:
I just tested this with the merge-devnet-3 config and it worked, so I think we're good to merge

paulhauner commented on consensus/proto_array/src/proto_array.rs:
```diff
@@ -142,11 +126,33 @@ impl ProtoArray {
 
+            // If we find the node for which the proposer boost was previously applied, decrease
+            // the delta by the previous score amount.
+            if self.previous_proposer_boost.root == node.root {
```
```suggestion
            if self.previous_proposer_boost.root != Hash256::zero()
                && self.previous_proposer_boost.root == node.root
            {
```

I'm not *really* sure this is necessary, but it prevents the scenario where a block with root `Hash256::zero()` will have its delta reduced without ever being increased (since the path to *increase* the boost is gated by `!= Hash256::zero()`).

It shouldn't happen in production, however we might see it during artificial testing. Also the spec likes to alias genesis to `Hash256::zero()` so perhaps this is useful defensively.

paulhauner commented on beacon_node/beacon_chain/src/schema_change/migration_schema_v7.rs:
```diff
@@ -0,0 +1,331 @@
+        .nodes
+        .iter()
+        .filter_map(|node| node.parent)
+        .collect::<Vec<_>>();
```
Collecting into a `HashSet` here would mean the later `nodes_referenced_as_parents::contains` is an `O(1)` search rather than `O(n)`.

Using a `Vec` shouldn't be *too* bad, but it's (almost) a 1-liner to use a `HashSet` instead.

paulhauner commented on beacon_node/beacon_chain/src/schema_change/migration_schema_v7.rs:
```diff
@@ -0,0 +1,331 @@
+        .core_proto_array()
+        .nodes
+        .iter()
+        .find_map(|node| {
```
In the case where there are multiple, distinct `justified_checkpoints` with a matching `finalized_checkpoint`, using a (forward) `iter` with a `find_map` is likely to discover the *lowest* `justified_checkpoint`.

Perhaps it would be ideal to replace this `find_map` with a `filter_map` and then reduce that set to a single element via [`Iterator::max_by_key`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.max_by_key). That way, we can find the `justified_checkpoint` with the *highest* epoch, which is the value we'd typically expect to see in the store.

realbigsean replied:
True, nice catch!

paulhauner commented on beacon_node/beacon_chain/src/schema_change/migration_schema_v7.rs:
```diff
@@ -0,0 +1,331 @@
+            let (justified_epoch, finalized_epoch) = nodes_v6
+                .get(index)
+                .map(|node: &ProtoNodeV6| (node.justified_epoch, node.finalized_epoch))
+                .ok_or_else(|| "Head index not found in legacy proto nodes".to_string())?;
```
```suggestion
                .ok_or_else(|| "Index not found in legacy proto nodes".to_string())?;
```

Sometimes this function will be supplied indices of ancestors of the head.

paulhauner commented on beacon_node/beacon_chain/src/schema_change/migration_schema_v7.rs:
```diff
@@ -0,0 +1,331 @@
+            let (justified_epoch, finalized_epoch) = nodes_v6
+                .get(index)
+                .map(|node: &ProtoNodeV6| (node.justified_epoch, node.finalized_epoch))
+                .ok_or_else(|| "Head index not found in legacy proto nodes".to_string())?;
```
```suggestion
                .ok_or_else(|| "Index not found in legacy proto nodes".to_string())?;
```

As above.

paulhauner commented on beacon_node/beacon_chain/src/schema_change/migration_schema_v7.rs:
```diff
@@ -0,0 +1,331 @@
+    for head in heads {
+        // `relevant_epochs` are epochs for which we will need to find the root at the start slot.
+        // We don't need to worry about whether the are finalized or justified epochs.
+        let mut relevant_epochs = vec![];
```
There would be no need to do a dedup later on if this were a `HashSet` rather than a `Vec`. Not a big deal at all, just something to consider.

paulhauner reviewed (changes requested):
This is looking really good! I really liked the used of `superstruct` and `derivative` to make things cleaner. This is critical code and your work really does justice to that.

I spent a couple of hours on this and I only have a handful of minor suggestions. I'm looking forward to seeing this merged!

realbigsean commented:
> I really liked the used of superstruct and derivative to make things cleaner.

Props to @michaelsproul for this!

realbigsean committed:
Update beacon_node/beacon_chain/src/schema_change/migration_schema_v7.rs

Co-authored-by: Paul Hauner <paul@paulhauner.com>

realbigsean committed:
Update beacon_node/beacon_chain/src/schema_change/migration_schema_v7.rs

Co-authored-by: Paul Hauner <paul@paulhauner.com>

realbigsean committed:
Update consensus/proto_array/src/proto_array.rs

Co-authored-by: Paul Hauner <paul@paulhauner.com>

realbigsean committed:
- Use `max_by_key` in `update_store_justified_checkpoint`
- Use a `HasSet` rather than a `Vec` in `map_relevant_epochs_to_roots`

realbigsean committed:
Merge branch 'justified-checkpoint-root' of https://github.com/realbigsean/lighthouse into justified-checkpoint-root

# Conflicts:
#	beacon_node/beacon_chain/src/schema_change/migration_schema_v7.rs

realbigsean committed:
comment update

realbigsean commented:
Ok all comments addressed! Diff here: https://github.com/sigp/lighthouse/compare/e266fe1539760faaaefbdcb8ddcd4f392207ec77..9d4692f6c

michaelsproul reviewed (approved):
Merge time!

bors r+

paulhauner reviewed (approved):
:rocket:
